presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -1104,6 +1107,32 @@ public PlanWithProperties visitUnion(UnionNode node, Context context)
                 }
             }
 
+            // parent does not have preference or prefers some partitioning without any explicit partitioning - just use
+            // children partitioning and don't GATHER partitioned inputs
+            if (!parentGlobal.isPresent() || parentGlobal.get().isDistributed()) {
+                if (!partitionedChildren.isEmpty() && unpartitionedChildren.isEmpty()) {
+                    if (!hasMultipleSources(plannedChildren.stream()
+                            .map(PlanWithProperties::getNode)
```

Comment:
Instead of mapping, you could just stream partitionedChildren instead. You've already checked that unpartitionedChildren is empty.